sycl/test/on-device/xocc/simple_tests/evaluation_tree.cpp

```diff
@@ -0,0 +1,188 @@
+          return dev_visit(
+              [&](auto&&... Others) {
+                std::forward<Func>(f)(
+                    std::forward<decltype(First)>(First),
+                    std::forward<decltype(Others)>(Others)...);
+              },
+              std::forward<Rest>(rest)...);
+        },
+        std::forward<Var>(var));
+}
+
+/// This is a relative pointer behaving mostly like T*. It doesn't manage the
+/// lifetime of the data pointed to. Also this class have shalow const semantic
+/// meaning that const rel_ptr<T> is equivalent to T* const not const T *. This
+/// pointer should behave like T* for any operation except memcpy or bitcasting
+/// the representation of the pointer. When memcpy'ed to an other place the
+/// offset between the pointer's address and the data it is pointed to will stay
+/// constant. this means that if the data is copied with the same memcpy the
+/// pointer is still pointing to the memcpy'ed data.
+/// This pointer can also be used to store pointers into files.
+template <typename T, typename IntTy = int64_t> class rel_ptr {
+  IntTy offset;
+
+  char *get_this() const { return ((char *)this); }
+  IntTy get_offset(T *t) const { return (((char *)t) - get_this()); }
+
+public:
+  T *get() const { return (T *)(get_this() + offset); }
+  T &operator*() const { return *get(); }
+  T *operator->() const { return get(); }
+  T &operator[](size_t e) const { return get()[e]; }
+  explicit operator bool() const { return get(); }
+  bool operator!() const { return !get(); }
+
+  rel_ptr() : rel_ptr(nullptr) {}
+  rel_ptr(T *t) : offset(get_offset(t)) {}
+  rel_ptr(const rel_ptr &other) : rel_ptr(other.get()) {}
+  rel_ptr &operator=(const rel_ptr &other) {
+    offset = get_offset(other.get());
+    return *this;
+  }
+
+  rel_ptr operator+(IntTy off) { return rel_ptr(get() + off); }
+  rel_ptr operator-(IntTy off) { return rel_ptr(get() - off); }
+  IntTy operator-(const rel_ptr &other) const { return offset - other.offset; }
+
+  bool operator==(const rel_ptr &other) const { return get() == other.get(); }
+  bool operator!=(const rel_ptr &other) const { return !(*this == other); }
+  friend bool operator==(const rel_ptr &p1, T *p2) { return p1.get() == p2; }
+  friend bool operator!=(const rel_ptr &p1, T *p2) { return !(p1 == p2); }
+
+  rel_ptr &operator++() {
+    *this = (*this) + 1;
+    return *this;
+  }
+  rel_ptr operator++(int) {
+    rel_ptr tmp = *this;
+    ++*this;
+    return tmp;
+  }
+  rel_ptr &operator--() {
+    *this = (*this) - 1;
+    return *this;
+  }
+  rel_ptr operator--(int) {
+    rel_ptr tmp = *this;
+    --*this;
+    return tmp;
+  }
+};
+
+/// Basic compute graph run in SYCL
+
+struct Constant {
+  int d;
+  int compute() { return d; }
+};
+
+struct node;
+
+struct BinOp { rel_ptr<node> elem[2]; };
+struct AddOp : BinOp { int compute(); };
+struct MulOp : BinOp { int compute(); };
+
+struct node {
+  std::variant<Constant, AddOp, MulOp> n;
+  int cache;
```

Comment:
The implentation is somehow broken since you do not really use a cache and so only works with a topologically sorted DAG.
So why not really have `std::optional<int> cache` instead and move `compute()` in `get_value()` instead?
But here we need to remove a recursion...
So perhaps a more generic way would be with no recursion to just return the optional with a value only if the value is there or compute the value if the operands have already a value... Cf infra

Reply:
Actually implementing `compute()` doing the computation if not done already is still useful.